src/libpriv/rpmostree-rpm-util.c

```diff
@@ -1128,3 +1097,52 @@ rpmostree_sack_get_by_pkgname (DnfSack     *sack,
   *out_pkg = g_steal_pointer (&ret_pkg);
   return TRUE;
 }
+
+GPtrArray*
+rpmostree_sack_get_packages (DnfSack *sack)
+{
+  hy_autoquery HyQuery query = hy_query_create (sack);
+  hy_query_filter (query, HY_PKG_REPONAME, HY_EQ, HY_SYSTEM_REPO_NAME);
+  return hy_query_run (query);
+}
+
+GPtrArray*
+rpmostree_sack_get_sorted_packages (DnfSack *sack)
+{
+  g_autoptr(GPtrArray) pkglist = rpmostree_sack_get_packages (sack);
+  g_ptr_array_sort (pkglist, (GCompareFunc)pkg_array_compare);
+  return g_steal_pointer (&pkglist);
+}
+
+gboolean
+rpmostree_create_rpmdb_pkglist_variant (int              rootfs_dfd,
+                                        GVariant       **out_variant,
+                                        GCancellable    *cancellable,
+                                        GError         **error)
+{
+  g_autoptr(RpmOstreeRefSack) refsack =
+    rpmostree_get_refsack_for_root (rootfs_dfd, ".", error);
+  if (!refsack)
+    return FALSE;
+
+  /* we insert it sorted here so it can efficiently be searched on retrieval */
+  g_autoptr(GPtrArray) pkglist = rpmostree_sack_get_sorted_packages (refsack->sack);
+
+  GVariantBuilder pkglist_v_builder;
+  g_variant_builder_init (&pkglist_v_builder, (GVariantType*)"a(stsss)");
+
+  const guint n = pkglist->len;
+  for (guint i = 0; i < n; i++)
+    {
+      DnfPackage *pkg = pkglist->pdata[i];
+      g_variant_builder_add (&pkglist_v_builder, "(stsss)",
+                             dnf_package_get_name (pkg),
+                             dnf_package_get_epoch (pkg),
+                             dnf_package_get_version (pkg),
+                             dnf_package_get_release (pkg),
+                             dnf_package_get_arch (pkg));
+    }
+
+  *out_variant = g_variant_builder_end (&pkglist_v_builder);
```

Comment:
Let's have a convention that `**out_variant` returns are *never* floating?  I.e. add `g_variant_ref_sink()` here.

Reply:
Yeah, agreed. It's just easier to think about. Done in ⬇️!